onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.cc

```diff
@@ -0,0 +1,217 @@
+          ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kXnnpackExecutionProvider, kMSInternalNHWCDomain, Start, End, Op)>
+
+#define KERNEL_CREATE_INFO(Start, Op)                                                      \
+  BuildKernelCreateInfo<                                                                   \
+      ONNX_OPERATOR_KERNEL_CLASS_NAME(kXnnpackExecutionProvider, kOnnxDomain, Start, Op)>, \
+      BuildKernelCreateInfo<                                                               \
+          ONNX_OPERATOR_KERNEL_CLASS_NAME(kXnnpackExecutionProvider, kMSInternalNHWCDomain, Start, Op)>
+
+VERSIONED_KERNEL_CLASS_NAME(1, 10, Conv);
+KERNEL_CLASS_NAME(11, Conv);
+
+VERSIONED_KERNEL_CLASS_NAME(1, 7, MaxPool);
+VERSIONED_KERNEL_CLASS_NAME(8, 9, MaxPool);
+VERSIONED_KERNEL_CLASS_NAME(10, 10, MaxPool);
+VERSIONED_KERNEL_CLASS_NAME(11, 11, MaxPool);
+KERNEL_CLASS_NAME(12, MaxPool);
+
+std::unique_ptr<KernelRegistry> RegisterKernels() {
+  auto kernel_registry = std::make_unique<onnxruntime::KernelRegistry>();
+
+  static const BuildKernelCreateInfoFn function_table[] = {
+      BuildKernelCreateInfo<void>,  // default entry to avoid the list becoming empty after ops-reducing
+      KERNEL_CREATE_INFO_VERSIONED(1, 10, Conv),
+      KERNEL_CREATE_INFO(11, Conv),
+
+      KERNEL_CREATE_INFO_VERSIONED(1, 7, MaxPool),
+      KERNEL_CREATE_INFO_VERSIONED(8, 9, MaxPool),
+      KERNEL_CREATE_INFO_VERSIONED(10, 10, MaxPool),
+      KERNEL_CREATE_INFO_VERSIONED(11, 11, MaxPool),
+      KERNEL_CREATE_INFO(12, MaxPool),
+  };
+
+  for (auto& function_table_entry : function_table) {
+    KernelCreateInfo info = function_table_entry();
+    if (info.kernel_def != nullptr) {  // filter disabled entries where type is void
+      ORT_THROW_IF_ERROR(kernel_registry->Register(std::move(info)));
+    }
+  }
+
+  return kernel_registry;
+}
+
+}  // namespace xnnpack
+
+using namespace xnnpack;
+
+XnnpackExecutionProvider::XnnpackExecutionProvider(const XnnpackExecutionProviderInfo& info)
+    : IExecutionProvider{kXnnpackExecutionProvider, true},
+      session_options_{info.session_options} {
+  // TODO: Could/should we provide our default CPU allocator to this call via an adapter?
+  xnn_status st = xnn_initialize(nullptr);
+  if (st != xnn_status_success) {
+    ORT_THROW("XNNPACK initialization failed with status ", st);
+  }
+
+  // TODO: Allocation planner calls GetAllocator for the individual EP. It would be better if it goes through
+  // the session state to get the allocator so it's per-device (or for the allocation planner to try the EP first
+  // and fall back to using session state next by passing in a functor it can use to call SessionState::GetAllocator).
+  // That way we only need one allocator per-device unless the EP needs/wants to override any previously registered
+  // allocator. Which also means one arena per-device if the per-EP allocators are using arenas.
```

Comment:
I'm not quite understand this part. Is it possible to avoid creating a new OrtMemoryInfo type? Can we let it share the same allocator as the CPU EP? How would it affect ORT's memory copy?

Reply:
There's a RegisterAllocator setup that is used to share an allocator between TRT and CUDA. I'll try and setup the same thing here in a separate PR as that needs some tweaks to work as (IMHO) expected.

Shouldn't affect copies as both XNNPACK and CPU EP are using the same device so no memcpy nodes should be being inserted.